training/MESSAGING_OVERVIEW.md

```diff
@@ -0,0 +1,138 @@
+
+To accomplish this, we use a command map structure in Chapel to map a string name to a Chapel function within a server Module. Adding a new module is simple. At the root level there is a `ServerModules.cfg` file that lists all Chapel Modules available for Command Map arguments. To add a new module, simply add the name of the module to this file. To learn how to structure a new Chapel Module, go to [Creating a New Module](#newModule).
+
+<a id="sendReq"></a>
+## How to Send a Request to the Server
+
+In general, a server request will look something like this:
+
+`repMsg = cast(str, generic_msg(cmd="serverFunction", args={"arg1": arg1, "arg2": arg2}))`
+
+We cast the return from `generic_msg` to a `str` in most cases to ensure our return is in a string format. The alternative is `memoryview`, which is a binary representation.
+
+To break this request down, `cmd` is the string name that maps to the function you want to use in Chapel. In most cases the string name will be almost identical to the Chapel function name to keep things easy to maintain. `args` is a JSON of all the arguments being passed to the server for processing. 
+
+<a id="receiveReq"></a>
+## Receiving a Request on the Server
+
+As mentioned above, the Command Map handles mapping the given `cmd` string to the proper function in the Chapel based on the available Chapel Modules. In order for a function within a module to be found by the Command Map it must first be registered. An example of this, using the request in the previous section as an example, would be as follows:
+
+`registerFunction("serverFunction", serverFunctionMsg, getModuleName());`
+
+The three components of the `registerFunction` method are the `cmd` string, the Chapel function name, and the Module name which in most cases we use `getModuleName()` to get the name of the module the function is called from.
+
+All messages sent to the server are in JSON format. So before any arguments sent from the client can be used, they must first be parsed. To do this use the function `parseMessageArgs(payload, argSize)`. This will parse your message arguments into a Map that can then be accessed using accessor methods. The most common of these accessor methods is `getValueOf(argName)` which will return the value of a argument that matches the passed in name.
+
+For example, assuming the server request in the above section is what is being received, there are two arguments in the JSON, `arg1` and `arg2`. To access these arguments in Chapel, we will parse the message, or `payload`, into a variable called `msgArgs`. Then using the `getValueOf()` accessor method, we will set variables `arg1` and `arg2` equal to the values of their corresponding JSON arguments.
+
+```chapel
+var msgArgs = parseMessageArgs(payload, argSize);
+var arg1 = msgArgs.getValueOf("arg1");
+var arg2 = msgArgs.getValueOf("arg2");
+```
+
+Now the values of `arg1` and `arg2` are available for use within this Chapel function.

```

Comment:
Can we go into a bit more detail on how to access arguments? This provides one case, but we really should highlight the functions that cast because they are used often.